runtime/js-compute-runtime/builtins/shared/console.cpp

```diff
@@ -431,22 +434,390 @@ static bool console_out(JSContext *cx, unsigned argc, JS::Value *vp) {
   return true;
 }
 
+// https://console.spec.whatwg.org/#assert
+// assert(condition, ...data)
+static bool assert(JSContext *cx, unsigned argc, JS::Value *vp) {
+  JS::CallArgs args = CallArgsFromVp(argc, vp);
+  auto condition = args.get(0).toBoolean();
+  // 1. If condition is true, return.
+  if (!condition) {
+    args.rval().setUndefined();
+    return true;
+  }
+
+  // 2. Let message be a string without any formatting specifiers indicating generically an
+  // assertion failure (such as "Assertion failed").
+  std::string message = "Assertion failed";
+  // 3. If data is empty, append message to data.
+  // 4. Otherwise:
+  // 4.1. Let first be data[0].
+  // 4.2. If Type(first) is not String, then prepend message to data.
+  // 4.3. Otherwise:
+  // 4.3.1. Let concat be the concatenation of message, U+003A (:), U+0020 SPACE, and first.
+  // 4.3.2. Set data[0] to concat.
+  auto length = args.length();
+  if (length > 1) {
+    message += ": ";
+    JS::RootedObjectVector visitedObjects(cx);
+    for (int i = 0; i < length; i++) {
+      JS::HandleValue arg = args.get(i);
+      std::string source = "";
+      auto result = ToSource(cx, source, arg, &visitedObjects);
+      if (result.isErr()) {
+        return false;
+      }
+      // strip quotes for direct string logs
+      if (source[0] == '"' && source[source.length() - 1] == '"') {
+        source = source.substr(1, source.length() - 2);
+      }
+      if (message.length()) {
+        message += " ";
+        message += source;
+      } else {
+        message += source;
+      }
```

Comment:
This pattern is repeated quite a lot, perhaps let's just refactor it into a separate function?